lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -6753,8 +6753,94 @@ bool isValidLLVMModule(Module *M, SPIRVErrorLog &ErrorLog) {
 }
 
+#if defined(LLVM_SPIRV_BACKEND_TARGET_PRESENT)
+namespace llvm {
+extern "C" bool
+SPIRVTranslateModule(Module *M, std::string &SpirvObj, std::string &ErrMsg,
+                     const std::vector<std::string> &AllowExtNames,
+                     const std::vector<std::string> &Opts);
+} // namespace llvm
+#endif
+
 namespace {
 
+#if defined(LLVM_SPIRV_BACKEND_TARGET_PRESENT)
+static inline Triple::SubArchType spirvVersionToSubArch(VersionNumber VN) {
+  switch (VN) {
+  case VersionNumber::SPIRV_1_0:
+    return Triple::SPIRVSubArch_v10;
+  case VersionNumber::SPIRV_1_1:
+    return Triple::SPIRVSubArch_v11;
+  case VersionNumber::SPIRV_1_2:
+    return Triple::SPIRVSubArch_v12;
+  case VersionNumber::SPIRV_1_3:
+    return Triple::SPIRVSubArch_v13;
+  case VersionNumber::SPIRV_1_4:
+    return Triple::SPIRVSubArch_v14;
+  case VersionNumber::SPIRV_1_5:
+    return Triple::SPIRVSubArch_v15;
+  case VersionNumber::SPIRV_1_6:
+    return Triple::SPIRVSubArch_v16;
+  }
+  return Triple::NoSubArch;
+}
+
+bool runSpirvBackend(Module *M, std::ostream *OS, std::string &ErrMsg,
+                     const SPIRV::TranslatorOpts &TranslatorOpts) {
+  static const std::set<ExtensionID> ImplementedBySpirvBE{
+      SPIRV::ExtensionID::SPV_EXT_shader_atomic_float_add,
+      SPIRV::ExtensionID::SPV_EXT_shader_atomic_float16_add,
+      SPIRV::ExtensionID::SPV_EXT_shader_atomic_float_min_max,
+      SPIRV::ExtensionID::SPV_INTEL_arbitrary_precision_integers,
+      SPIRV::ExtensionID::SPV_INTEL_cache_controls,
+      SPIRV::ExtensionID::SPV_INTEL_global_variable_fpga_decorations,
+      SPIRV::ExtensionID::SPV_INTEL_global_variable_host_access,
+      SPIRV::ExtensionID::SPV_INTEL_optnone,
+      SPIRV::ExtensionID::SPV_INTEL_usm_storage_classes,
+      SPIRV::ExtensionID::SPV_INTEL_subgroups,
+      SPIRV::ExtensionID::SPV_KHR_uniform_group_instructions,
+      SPIRV::ExtensionID::SPV_KHR_no_integer_wrap_decoration,
+      SPIRV::ExtensionID::SPV_KHR_float_controls,
+      SPIRV::ExtensionID::SPV_KHR_expect_assume,
+      SPIRV::ExtensionID::SPV_KHR_bit_instructions,
+      SPIRV::ExtensionID::SPV_KHR_linkonce_odr,
+      SPIRV::ExtensionID::SPV_INTEL_inline_assembly,
+      SPIRV::ExtensionID::SPV_INTEL_bfloat16_conversion,
+      SPIRV::ExtensionID::SPV_KHR_subgroup_rotate,
+      SPIRV::ExtensionID::SPV_INTEL_variable_length_array,
+      SPIRV::ExtensionID::SPV_INTEL_function_pointers,
+      SPIRV::ExtensionID::SPV_KHR_shader_clock,
+      SPIRV::ExtensionID::SPV_KHR_cooperative_matrix,
+      SPIRV::ExtensionID::SPV_KHR_non_semantic_info};
+  static const std::string DefaultTriple = "spirv64-unknown-unknown";
+  static const std::vector<std::string> Opts{"--avoid-spirv-capabilities",
+                                             "--translator-compatibility-mode"};
+
+  std::function<bool(SPIRV::ExtensionID)> Filter =
+      [](SPIRV::ExtensionID Ext) -> bool {
+    return ImplementedBySpirvBE.find(Ext) != ImplementedBySpirvBE.end();
```

Comment:
> this table has to be updated
Yes, this is by intent. The rationale is that Translator is a client of SPIRV BE service in this case, and a set of extensions supported by SPIRV BE is a part of the contract. When the contract is updated on the service side (new extension is added), the client must update its part of the contract as well (allow to pass this extension to SPIRV BE).

From the technical point of view, this list must be filtered anyway, either by SPIRV BE, or by Translator, otherwise, SPIRV BE fails with the complaint about unknown extension. So this PR proposes an explicit way of resolving this, by clearly state the contract (limitations, in a form of supported extensions that Translator aware of). If decision were rather to implement this on SPIRV BE side by ignoring unknown extension, we would be unable to differ whether client passes a valid but yet unsupported extension, or it's a user error and a valid and supported extension is just misspelled.

Reply:
Lets say we have a call to `__spirv_MyBrandNewInstruction` in LLVM IR, which is recognizable by SPIR-V translator as a part of `MyBrandNew` SPIR-V extension. Right now, if we pass it to the translator with the enabled `MyBrandNew` extension, then the translator will generate `OpMyBrandNewInstruction`. If `MyBrandNew` is not enabled - the translator will report an error. Meanwhile, if the extension **is not known** to the translator - it won't generate an error and instead will leave this external function call untouched (may be to consider to report an error in this case actually).

What is the backend's behavior, if it sees `__spirv_MyBrandNewInstruction` function call, but knows nothing about `MyBrandNew` extension? If it reports an error - it's great and I have no concerns with filtering out the extensions. If it doesn't report an error and just generates the external function call - well, it matches the translator's behavior for unknown extensions, but it doesn't match the translator's behavior for this very extension, so instead of truthfully fail with an error, saying that the extension passed along with --spirv-ext option is unsupported by the backend, the backend will silently emit some function call, which later SPIR-V consumer has to resolve somehow, and this might be a problem.

Reply:
Yes, indeed new users/use cases would get just OpFunctionCall instead of an error, as you've described in Translator's behavior. However, consider existing users.

For example, the Triton compiler enables all extensions. This means that without filtering in Translator's side calls from the Triton compiler will always result in an error.

Another example is SYCL very (really) long list of extensions. Without filtering on Translator side SYCL users need to refactor their calls to be sure that SPIRV BE serves their request.

In other words, if we don't use filtering, we just move the problem deeper, closer to an end-user. If we filter in SPIRV BE we lose information/chance to report misspelling names. Filtering in Translator does not solve all the problems, as your example demonstrated, but it looks like a minimal of all alternative issues.

As a quick and naive resolution of the problem that you've mentioned, it would be easier to introduce an option in SPIRV BE that controls whether __spriv_XYZ has to be resolved to a known instruction, by the analogy with Translator's option to allow unknown intrinsics by prefix.

Reply:
> For example, the Triton compiler enables all extensions (just enableAllExtensions()). This means that without filtering in Translator's side calls from the Triton compiler will always result in an error.

Seems like not the best idea for me. Can't stop myself but wonder if they will report a 100% regression with the translator flow due to recent work related to opaque pointers extension started by Viktoria. Btw, is --spirv-ext=+all is being handled implicitly with this patch?

Reply:
This patch is just going through the list of extensions, after +all had already been parsed, so from the patch perspective it doesn't matter if the list was explicit or +all, the patch deals with parsed representation for both.